### PR TITLE
[Model] Boogu-Image: reference-image edit (TI2I / I2I) pipeline

### PR DESCRIPTION
**Pull request by NancyFyong**
> **Stacked on #33182** (its base branch), so the diff below still contains #33182's
> commits. Reviewable delta = `git diff feat/boogu-image...` → 9 files, +784/−15.
> Will rebase and drop the duplicates once #33182 lands.

## Motivation

#33182 added the text-to-image `BooguImagePipeline`. The same Boogu-Image checkpoint
also serves **reference-image editing** (TI2I / I2I). This PR extends that one pipeline
to serve both: a single pipeline + config (`task_type=TI2I`) handles text-only and
reference-conditioned requests, matching the one-pipeline house pattern (`GlmImage`,
`Flux2`). With no reference image it degrades to the existing T2I behavior, so base T2I
output is unchanged.

## Modifications

The edit path reuses the DiT, VAE, Qwen3-VL encoder, and scheduler unchanged. A
reference image is routed to **both** the Qwen3-VL encoder (image-aware instruction
embeddings) and the VAE (reference latents threaded through the DiT via
`ref_image_hidden_states`). With no reference image the pipeline degrades to T2I.

- `configs/pipeline_configs/boogu_image.py` — `BooguImagePipelineConfig` gains
  `task_type=TI2I` and `BooguImageCFGPolicy`: dual text+image CFG (1/2/3 branches),
  collapsing to standard CFG `c + (s-1)(c-u)` when there is no reference.
- `configs/sample/boogu_image.py` — adds `guidance_scale_2` (image guidance, default
  `1.0` = off, so a text-only request runs exactly two CFG branches).
- `runtime/pipelines/boogu_image_pipeline.py` — builds the reference-aware
  `BooguImageEncodingStage`, so the single pipeline serves both paths; no
  `--pipeline-class-name` and no separate edit pipeline.
- `.../stages/model_specific_stages/boogu_image.py` — dual-resolution reference
  preprocessing (384 px VLM copy, 2048 px VAE copy), VAE-encode `(sample - shift) * scale`,
  TI2I system prompt with image-before-text ordering; falls back to the plain T2I encode
  when no reference is present.
- `.../stages/model_specific_stages/boogu_image_processor.py` — native port of the
  reference downscale-only, alignment-aware preprocessing (drops the external `boogu` dep).
- `runtime/warmup_request_builder.py` — opts out of the synthetic warmup image so the
  server warms at the real 1024×1024, following the `GlmImagePipelineConfig` precedent.
- `runtime/pipelines_core/schedule_batch.py` — additive `ref_image_hidden_states` field.

Boogu's `head_dim` is 120, which is not an FA-supported size, so q/k/v are zero-padded to
128 and the output sliced back. The softmax scale stays `head_dim**-0.5` = 1/sqrt(120):
zero columns leave `q·k` unchanged, so the padded dim must not enter the temperature. It
is passed explicitly because `Attention.__init__` defaults to `head_size**-0.5`, and
`head_size` there is the padded 128. (This lives in the base T2I DiT from #33182; it is
not part of this PR's diff.)

## Accuracy Tests

CPU-only unit tests: `test_boogu_image.py` **14 passed**. They pin the dual-CFG branch
construction and combine formula, the T2I-unchanged path (reference-free build → 2
branches, combine → `c + (s-1)(c-u)`), `task_type=TI2I`, VAE-encode shift-then-scale
order, `guidance_scale_2` defaulting off, and image-before-text message building.

Same weights and conditioning across all three stacks: 1024×1024, 50 steps, seed 42,
text guidance 4.0, image guidance 1.0, empty negative prompt, bf16.

Reference input for the edit row (`input_image_examples/03.jpg`): ti2i_input_ref

| | diffusers (reference) | vllm-omni | sglang (this PR) |
|---|---|---|---|
| **TI2I** edit<br>`把背景替换到沙滩.` | ti2i_diffusers | ti2i_vllm_omni | ti2i_sglang |
| **T2I** base<br>*(decoupling check)* | t2i_diffusers | t2i_vllm_omni | t2i_sglang |

sglang's edit output reproduces the reference sample at **36.8 dB** vs diffusers; the
residual is bf16 / kernel drift, and `torch.compile` does not change it. Base T2I is
unaffected — **34.7 dB** vs diffusers, and folding the edit path into the shared pipeline
left the T2I output identical to the pre-merge run (**42.3 dB** pairwise, within bf16
run-to-run noise). vllm-omni's edit is a *different valid sample* rather than a degraded
one — lower horizon, brighter ocean — while its T2I matches, so that difference is
specific to its reference-image conditioning path.

## Speed Tests and Profiling

Full-pipeline wall time per image (text encode + 50 steps + VAE decode), 1024×1024,
concurrency 1, warmup excluded, H20. All four sglang arms are from the single merged
build.

| | diffusers | vllm-omni<br>eager | vllm-omni<br>compile | sglang<br>eager | sglang<br>compile |
|---|---|---|---|---|---|
| **TI2I** edit | 159.84 s | 136.76 s | 130.71 s | 143.97 s | **124.81 s (1.28×)** |
| **T2I** base | 68.21 s | 61.03 s | 58.10 s | 64.67 s | **54.77 s (1.25×)** |

sglang compile is fastest on both paths; eager-vs-eager it trails vllm-omni by ~5 %,
because compile buys sglang −13.3 %/−15.3 % vs vllm-omni's −4.4 %/−4.8 % (vllm-omni's
`_repeated_blocks` excludes its double-stream and refiner layers).

Head-dim padding is what makes FA usable here, and it is worth **−12.6 %** on TI2I
(SDPA 164.76 s → padded FA 144.02 s) and **−7.4 %** on T2I (69.84 s → 64.67 s), eager.

Nearly all the time is the denoise loop — TI2I e2e 143.97 s against a 143.39 s denoise
sum, so the edit path's extra VAE encode + image-fused VLM forward is ~0.6 s.

**Caveats:** every GPU on this host carries a foreign tenant at ~79–81 % utilization, so
absolute seconds are contended. diffusers and sglang ran on the same card (ratios are
apples-to-apples); vllm-omni ran on a second card, so its column is cross-card. The T2I
image cell and the T2I latency come from different runs, because vllm-omni's timed T2I
arm uses `dataset: random`, which substitutes its own prompt and injects a synthetic
negative prompt.

## Checklist

- [x] Format code with pre-commit
- [x] Add unit tests
- [ ] Update documentation (compatibility matrix — add the edit pipeline row)
- [x] Provide accuracy and speed benchmark results
- [x] Follow the SGLang code style guidance

<sub>AI assistance (Claude Code) was used for implementation and benchmarking; every line
reviewed, and all numbers above are from real runs on the hardware described.</sub>

---
### CI States

Latest PR Test (Base): :x: **Missing `run-ci` label** -- add it to run CI tests.
Latest PR Test (Extra): :x: **Blocked** -- `run-ci` is required first.
Latest PR Test (AMD ROCm 7.2): :heavy_minus_sign: **No AMD PR run found for this commit**.

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/boogu_image.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        batch.height = int(preprocessed.shape[-2])
+        batch.width = int(preprocessed.shape[-1])
+
+        scaling_factor, shift_factor = (
+            server_args.pipeline_config.get_decode_scale_and_shift(
+                device, None, self.vae
+            )
+        )
+        gen = batch.generator
+        vae_generator = gen if isinstance(gen, torch.Generator) else None
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/boogu_image.py]: `batch.generator` is always a list after input validation, so this passes `None` to the VAE sampler. The reference latent then ignores the request seed and may also differ across CFG-parallel ranks. Please unwrap the per-request generator here (`batch.generator[0]` for the current path).

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/boogu_image.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        )
+        return uses
+
+    @torch.no_grad()
+    def forward(self, batch: Req, server_args: ServerArgs) -> Req:
+        ref_images = batch.condition_image
+        if not ref_images:
+            return super().forward(batch, server_args)
+
+        raw_pil = ref_images[0].convert("RGB")
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/boogu_image.py:110]: We silently ignore every reference image after the first. Since Boogu currently supports one reference image, can we reject `len(ref_images) != 1` explicitly?
